### ISSUE TITLE
Resource checking should be resilient to a 'stuck' worker

### ISSUE THREAD

**Issue by vito** [concourse/concourse#3928]
## What challenge are you facing?

[Prod](https://ci.concourse-ci.org) had its `concourse` resource stop checking for over four days. Investigating the stack dump revealed that a lock was held open and the goroutine holding it was stuck on this line:

https://github.com/concourse/concourse/blob/8d6219d0e1d82ef897838d229dbe4394dd31ecef/atc/resource/run_script.go#L136

```
goroutine 1021 [chan receive, 5834 minutes]:
github.com/concourse/concourse/atc/resource.(*resource).runScript(0xc039940220, 0x2607d20, 0xc01888ae40, 0x1ced6e0, 0x13, 0x0, 0x0, 0x0, 0x1b073c0, 0xc03959f700, ...)
	/tmp/build/70f2e240/concourse/atc/resource/run_script.go:136 +0x724
github.com/concourse/concourse/atc/resource.(*resource).Check(0xc039940220, 0x2607d20, 0xc01888ae40, 0xc0043bc480, 0xc004e71710, 0xc03959f6f0, 0x38384c8, 0x25e54e0, 0xc02c816900, 0x1cd7271)
	/tmp/build/70f2e240/concourse/atc/resource/resource_check.go:17 +0x11c
github.com/concourse/concourse/atc/radar.(*resourceScanner).check(0xc000e2cc60, 0x2627140, 0xc018487920, 0x2656480, 0xc046598f20, 0x26347a0, 0xc035b89f20, 0xc004e71710, 0xc01100c000, 0x8, ...)
	/tmp/build/70f2e240/concourse/atc/radar/resource_scanner.go:360 +0xbeb
github.com/concourse/concourse/atc/radar.(*resourceScanner).scan(0xc000e2cc60, 0x2627140, 0xc018487920, 0x765, 0x0, 0x0, 0x0, 0x0, 0x0)
	/tmp/build/70f2e240/concourse/atc/radar/resource_scanner.go:251 +0x7d6
github.com/concourse/concourse/atc/radar.(*resourceScanner).Run(0xc000e2cc60, 0x2627140, 0xc000e8dec0, 0x765, 0xdf8475801, 0x25c7ce0, 0xc00022e940)
	/tmp/build/70f2e240/concourse/atc/radar/resource_scanner.go:63 +0x97
github.com/concourse/concourse/atc/radar.(*intervalRunner).Run(0xc000a36b40, 0x2607ca0, 0xc000cee800, 0x0, 0x0)
	/tmp/build/70f2e240/concourse/atc/radar/interval_runner.go:68 +0x2d6
github.com/concourse/concourse/atc/radar.(*Runner).scanResources.func1(0xc000c7dbd0, 0x2627140, 0xc000e8cd80, 0x2607ca0, 0xc000cee800, 0x2656480, 0xc0015998c0, 0xc0003e8200, 0x10)
	/tmp/build/70f2e240/concourse/atc/radar/runner.go:118 +0x11d
created by github.com/concourse/concourse/atc/radar.(*Runner).scanResources
	/tmp/build/70f2e240/concourse/atc/radar/runner.go:113 +0x336
```

(full stack available [here](https://gist.githubusercontent.com/pivotal-jwinters/1acbcd196230dc376aca1a1a8fe5c4e1/raw/4f1cf996accd59b9febca9a4a7e6951d8471246b/atc-stack-dump))

## What would make this better?

It looks like it got stuck waiting on the process to exit after enforcing the default resource check timeout (1 hour). While stuck in this way the lock is still being held.

We probably do this to ensure the goroutine calling `.Wait` eventually exits. In this case, though, the worker was pretty broken and seems to have never returned the exit status. Supposedly the worker was even re-created, so it's kind of odd that things got stuck instead of eventually erroring or something from the worker being destroyed.

1. I would like to start with reproducing the issue first. Can we simulate a 'stuck' worker and replicate this failure mode? Even better, can a TopGun test be written?
1. Can we enforce the timeout more cleanly? It would be nice if we can just give up client-side, release all client resources (goroutines, in-flight requests, etc.) and leave everything else up to the worker, if it's even capable of cleaning up its side of things. Right now the code will handle a legitimately hanging check *process* pretty cleanly, but if the problem is more fundamental our 'giving up' code...doesn't give up.
1. Should we shorten the default check timeout? 1 hour is super long. :thinking:

**Comment by vito** [comment]
Somewhat related: https://github.com/concourse/concourse/issues/1954

@ddadlani Dragging this into the Runtime backlog while it's fresh in our minds.

**Comment by marco-m** [comment]
As an additional data point, we also saw this happening sometimes (we saw the same symptom: resource checking stuck; cannot say if it was the same cause).

**Comment by clarafu** [comment]
We also saw the resource checks hanging happen in our Wings environment with the following stack dump
```
net/http.(*persistConn).roundTrip(0xc02d685d40, 0xc00206b8c0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:2174 +0x75c
net/http.(*Transport).roundTrip(0xc026488360, 0xc04fd0d000, 0x1cc4700, 0x1, 0xc04fd0d000)
	/usr/local/go/src/net/http/transport.go:481 +0xa1b
net/http.(*Transport).RoundTrip(0xc026488360, 0xc04fd0d000, 0xc000f955a0, 0xcee00, 0x19696a0)
	/usr/local/go/src/net/http/roundtrip.go:17 +0x35
github.com/concourse/concourse/atc/worker/transport.(*gardenRoundTripper).RoundTrip(0xc03384ec40, 0xc04fd0cf00, 0x5cfeffb9, 0xc02bc4a2a0, 0x48d796)
	/tmp/build/70f2e240/concourse/atc/worker/transport/garden_round_tripper.go:48 +0x135
github.com/concourse/retryhttp.(*RetryRoundTripper).RoundTrip.func1(0xc0267a6060, 0xc000f955a0)
	/tmp/build/70f2e240/gopath/pkg/mod/github.com/concourse/retryhttp@v1.0.1/retry_round_tripper.go:56 +0x92
github.com/cenkalti/backoff.RetryNotify(0xc0044a7040, 0x2577b20, 0xc0267a6060, 0x0, 0x0, 0x0)
	/tmp/build/70f2e240/gopath/pkg/mod/github.com/cenkalti/backoff@v2.1.1+incompatible/retry.go:37 +0xb9
github.com/cenkalti/backoff.Retry(...)
	/tmp/build/70f2e240/gopath/pkg/mod/github.com/cenkalti/backoff@v2.1.1+incompatible/retry.go:24
github.com/concourse/retryhttp.(*RetryRoundTripper).RoundTrip(0xc03384ec80, 0xc04fd0cf00, 0xc03384ec80, 0x0, 0x0)
	/tmp/build/70f2e240/gopath/pkg/mod/github.com/concourse/retryhttp@v1.0.1/retry_round_tripper.go:55 +0x1ce
net/http.send(0xc04fd0cf00, 0x2563280, 0xc03384ec80, 0x0, 0x0, 0x0, 0xc02e016d28, 0xc00be4dc00, 0x1, 0x0)
	/usr/local/go/src/net/http/client.go:250 +0x461
net/http.(*Client).send(0xc00e840420, 0xc04fd0cf00, 0x0, 0x0, 0x0, 0xc02e016d28, 0x0, 0x1, 0x20)
	/usr/local/go/src/net/http/client.go:174 +0xfb
net/http.(*Client).do(0xc00e840420, 0xc04fd0cf00, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/client.go:641 +0x279
net/http.(*Client).Do(...)
	/usr/local/go/src/net/http/client.go:509
github.com/concourse/concourse/atc/worker/transport.(*WorkerHijackStreamer).Stream(0xc00e8404e0, 0x1d00e18, 0x6, 0x2561760, 0xc00206b6b0, 0x0, 0x0, 0x1d1195e, 0x10, 0x0, ...)
	/tmp/build/70f2e240/concourse/atc/worker/transport/hijack_streamer.go:57 +0xf6
code.cloudfoundry.org/garden/client/connection.(*connection).do(0xc013efe8c0, 0x1d00e18, 0x6, 0x1c82ec0, 0xc02d852780, 0x1992120, 0xc02d93cc20, 0x0, 0x0, 0x0, ...)
	/tmp/build/70f2e240/gopath/pkg/mod/code.cloudfoundry.org/garden@v0.0.0-20181108172608-62470dc86365/client/connection/connection.go:587 +0xd4
code.cloudfoundry.org/garden/client/connection.(*connection).Create(0xc013efe8c0, 0xc0264261b0, 0x24, 0x0, 0xc003204360, 0x59, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/build/70f2e240/gopath/pkg/mod/code.cloudfoundry.org/garden@v0.0.0-20181108172608-62470dc86365/client/connection/connection.go:130 +0xde
code.cloudfoundry.org/garden/client.(*client).Create(0xc041b0e9d0, 0xc0264261b0, 0x24, 0x0, 0xc003204360, 0x59, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/build/70f2e240/gopath/pkg/mod/code.cloudfoundry.org/garden@v0.0.0-20181108172608-62470dc86365/client/client.go:31 +0x6b
github.com/concourse/concourse/atc/worker.workerHelper.createGardenContainer(0x7fe900ef6f50, 0xc041b0e9d0, 0x25c9040, 0xc02ac1a700, 0x25df8c0, 0xc0003e7350, 0x25aa1a0, 0xc0009d4260, 0x25e8900, 0xc0457d67e0, ...)
	/tmp/build/70f2e240/concourse/atc/worker/worker_helper.go:92 +0x45a
github.com/concourse/concourse/atc/worker.(*gardenWorker).FindOrCreateContainer(0xc0261d95e0, 0x25a20a0, 0xc0000b0010, 0x25c1840, 0xc005f31ec0, 0x25999a0, 0x37e78b0, 0x257f6e0, 0xc031a5a820, 0x1d001c5, ...)
	/tmp/build/70f2e240/concourse/atc/worker/worker.go:239 +0x82b
github.com/concourse/concourse/atc/radar.(*resourceScanner).check(0xc00195a510, 0x25c1840, 0xc005f31ec0, 0x25f1460, 0xc0247ede40, 0x25cf4a0, 0xc022585ce0, 0xc043a25f50, 0x0, 0x0, ...)
	/tmp/build/70f2e240/concourse/atc/radar/resource_scanner.go:334 +0x8d5
github.com/concourse/concourse/atc/radar.(*resourceScanner).scan(0xc00195a510, 0x25c1840, 0xc005f31ec0, 0x8d440, 0x0, 0x0, 0x0, 0x0, 0x0)
	/tmp/build/70f2e240/concourse/atc/radar/resource_scanner.go:251 +0x7d6
github.com/concourse/concourse/atc/radar.(*resourceScanner).Run(0xc00195a510, 0x25c1840, 0xc001296960, 0x8d440, 0xdf8475801, 0x0, 0x0)
	/tmp/build/70f2e240/concourse/atc/radar/resource_scanner.go:63 +0x97
github.com/concourse/concourse/atc/radar.(*intervalRunner).Run(0xc0010239f0, 0x25a2060, 0xc001917280, 0x0, 0x0)
	/tmp/build/70f2e240/concourse/atc/radar/interval_runner.go:68 +0x2d6
github.com/concourse/concourse/atc/radar.(*Runner).scanResources.func1(0xc000fe0d20, 0x25c1840, 0xc000554060, 0x25a2060, 0xc001917280, 0x25f1460, 0xc000b9c160, 0xc000cee980, 0x1c)
	/tmp/build/70f2e240/concourse/atc/radar/runner.go:118 +0x11d
created by github.com/concourse/concourse/atc/radar.(*Runner).scanResources
	/tmp/build/70f2e240/concourse/atc/radar/runner.go:113 +0x336
```

It seems to be stuck in the FindOrCreateContainer, maybe we should adding a timeout when we try talk to garden?

**Comment by ddadlani** [comment]
After discussion with @clarafu, the issue is occurring because the ATC waits on check container creation to return and in the case where the worker disappears, the ATC keeps waiting.

We could probably extend this to add a `ctx.WithTimeout()` to both container and volume creation/streaming.

**Comment by kcmannem** [comment]
looks like we gotta pass the context to the actual request object. This is done in the garden client which is not owned by us. Exact location is here 

https://github.com/cloudfoundry/garden/blob/e9c503bfdeedec3e591f46a85e3cd0e7b1ebecc1/client/connection/connection_hijacker.go#L48 
and here 
https://github.com/cloudfoundry/garden/blob/e9c503bfdeedec3e591f46a85e3cd0e7b1ebecc1/client/connection/connection_hijacker.go#L96 

We either fork it or make our own garden client and propagate the context everywhere... 😢

**Comment by cirocosta** [comment]
@kcmannem supporting cancellation through `context` is exactly why I implemented a custom single-method garden client in the "worker healthchecking" PR 😅 It'd be *great* if we had `context` passing around there in `cloudfoundry/garden`'s own code 🙌

**Comment by kcmannem** [comment]
if we're going to make the switch to containerd, i'd wait for that to happen since its client is built properly with contexts. https://github.com/containerd/containerd/blob/master/client.go

**Comment by clarafu** [comment]
This would also help with not only resource checks hanging after a worker goes missing, but also for get and put containers. We saw within the algorithm environment, when a worker needs to be restarted because of some filesystem failures, builds would be stuck in running state because of the get containers stuck waiting on garden.

**Comment by ddadlani** [comment]
@clarafu That only happens because of the lock that is held. We might be able to get away with using the context.WithTimeout to cancel the `FindOrCreateContainer` in the ATC worker, and not bubble the context cancellation down to garden. 

1) If the ATC goes down before calling `createGardenContainer`, no problems. 
2) If it goes down after making the call,
  a) the call succeeded, so the container will exist when we try to find it the next time
  b) the call failed, so we won't find it on the worker and will create it when we try again

We'll need to investigate to make sure that everything that is cancelled because of a timeout does actually get retried, and that there are no orphaned containers left around. But it would make this story a lot smaller and easier to implement if we limit it to the ATC.

**Comment by kcmannem** [comment]
talked to garden folks, they're open to us making a pr for contexts in the garden client.

**Comment by kcmannem** [comment]
I'm going to try to do what @ddadlani suggested first though cause its less painful

**Comment by kcmannem** [comment]
Looks like we have to fork garden. The garden client interface is embedded in the `Backend interface{}` https://github.com/cloudfoundry/garden/blob/master/backend.go#L7-L14 so any changes to the client methods would have to be done to all the garden backends. This is not doable.

**Comment by ddadlani** [comment]
Quick update: The garden team would like us to PR the context changes all the way through to `guardian` and `greenhouse` as well. This makes sense, but it is a large chunk of work and since we see this issue occurring mostly during worker recreates, it makes more sense for us to find a workaround.

We experimented with creating our own version of the `garden.Client`, called `gClient`, in our own code base, which passes a `context.WithTimeout()` into its requests. This allows us to cancel the request from our end.

However, there are two different ways we see this pending check issue, as identified by the two stack traces shared above:

1) The first one is in this line of code: https://github.com/concourse/concourse/blob/82dac8d08a215231f875dfaa6985809649dad6bd/atc/worker/worker_helper.go#L49

This occurs when the worker dies (or is recreated) while the check container is being created. By passing the context into this garden client request, we can cancel the request and ensure that the check lock is freed. We run the risk of a check container still existing on the worker, but since this happens most often during a worker recreate, the risk is minimal.

2) The second way is caused by this line of code: 
https://github.com/concourse/concourse/blob/8d6219d0e1d82ef897838d229dbe4394dd31ecef/atc/resource/run_script.go#L136

The call to `container.Stop` is asynchronous, so we wait for `processExited` as a way to ensure that the container truly exited. 

However if [`process.Wait()`](https://github.com/concourse/concourse/blob/8d6219d0e1d82ef897838d229dbe4394dd31ecef/atc/resource/run_script.go#L105) never returns, `processExited` won't get signaled.
We are currently investigating the possibility that our fix for #3310 also caused `process.Wait()` to never return.

**Comment by xtreme-sameer-vohra** [comment]
**This is WIP**
List of interface methods that interact with Worker during a  [Resource Scanner check](https://github.com/concourse/concourse/blob/master/atc/radar/resource_scanner.go#L266) that will need a `context` passed in
- [ ] worker.FindOrCreateContainer
  -  [ ] gardenClient.Lookup
     - [ ] gardenClient.Connection.List
    
     - [ ] volumeClient.FindOrCreateVolumeForContainer
       - [ ] volumeClient.LookupVolume 
- [ ] resource.Check

**Comment by kcmannem** [comment]
@xtreme-sameer-vohra 

heres the test we can use to validate the context expiry. No need to do a full on integration test.

```go
	Describe("Context Timeout", func(){
		Context("passing a ctx with Timeout should cancel the client request when garden server does not respond", func() {
			go func() {
				http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
					for { time.Sleep(1*time.Second) }
				})

				http.ListenAndServe(":8989", nil)
			}()

			ctx, _ := context.WithTimeout(context.Background(), 15*time.Second)
			fakeTransport := new(transportfakes.FakeTransportDB)
			fakeLogger := new(lagertest.TestLogger)
			hostname := new(string)
			*hostname = "localhost:8989"
			clientFactory := gclient.NewGardenClientFactory(
				fakeTransport,
				fakeLogger,
				"wont-talk-to-you",
				hostname,
				retryhttp.NewExponentialBackOffFactory(1*time.Second),
			)

			client := clientFactory.NewClient()
			_, err := client.Create(ctx, garden.ContainerSpec{})
			Expect(err).To(HaveOccurred())
		})
	})
```

**Comment by cirocosta** [comment]
@kcmannem that seems quite similar to how it's done for healthchecking 😁  

https://github.com/concourse/concourse/blob/6360a0454a4f56fca14f95a5b91542a5c898b795/worker/healthcheck/garden_test.go#L51-L61

https://github.com/concourse/concourse/blob/6360a0454a4f56fca14f95a5b91542a5c898b795/worker/healthcheck/garden_test.go#L33-L37

**Comment by kcmannem** [comment]
@cirocosta oooo didn't know ginkgo provided servers. Thanks for the reference.

**Comment by xtreme-sameer-vohra** [comment]
Here is some context on what we learnt as part of this story

Garden Client has 2 mediums to communicate with the Garden server as part of the [HijackStreamer Interface](https://github.com/concourse/concourse/blob/d15e2ed8d257ff9d6aa82a8fd401f94799278468/atc/worker/gclient/connection/connection.go#L75-L78)
- Stream
- Hijack ( this is not the same as `hijack` in the context of Concourse. It is `hijack`ing the underlying net.Conn from an httpClient )

`Stream` in the HijackStreamer is used for most requests that are short in nature. We decided to add a static 5 min overall request timeout for requests using this medium. This affects more than just `worker.FindOrCreateContainer` used by `resource_scanner`. However, we believe 5 mins is a very conservative value for these requests and if the request hasn't completed in 5 minutes, something is wrong.

For `Hijack`, a ctx is passed down through the [Container Interface](https://github.com/concourse/concourse/blob/d15e2ed8d257ff9d6aa82a8fd401f94799278468/atc/worker/gclient/container.go#L102-L108) allowing the request to be cancelled. This addresses the primary stuck stack trace raised above during [context.Done()](https://github.com/concourse/concourse/blob/8d6219d0e1d82ef897838d229dbe4394dd31ecef/atc/resource/run_script.go#L136)

On a `Container.Attach`, we observed upto 3 net.Conn being created for `stdin`, `stdout` & `stderr`. While a `container.Stop` method might succeed, the 3 net.Conn's need to be closed from the server side for the client to disconnect cleanly. Our change, passes down a `ctx` allowing the net.conns to be closed off on the client side.
